src/sidebar/services/test/groups-test.js

```diff
@@ -819,43 +820,46 @@ describe('groups', function () {
     });
 
     context('when a new frame connects', () => {
-      it('should refetch groups if main frame URL has changed', () => {
+      it('should refetch groups if main frame URL has changed', async () => {
         const svc = service();
 
         fakeStore.setState({
           frames: [{ uri: 'https://domain.com/page-a' }],
         });
-        return svc
-          .load()
-          .then(() => {
-            // Simulate main frame URL change, eg. due to client-side navigation in
-            // a single page application.
-            fakeApi.groups.list.resetHistory();
-            fakeStore.setState({
-              frames: [{ uri: 'https://domain.com/page-b' }],
-            });
-
-            return fakeRootScope.eventCallbacks[events.FRAME_CONNECTED]();
-          })
-          .then(() => {
-            assert.calledOnce(fakeApi.groups.list);
-          });
+        await svc.load();
+
+        // Simulate main frame URL change, eg. due to client-side navigation in
+        // a single page application.
+        fakeApi.groups.list.resetHistory();
+        fakeStore.setState({
+          frames: [{ uri: 'https://domain.com/page-b' }],
+        });
+
+        await waitFor(() => fakeApi.groups.list.callCount > 0);
+        assert.calledOnce(fakeApi.groups.list);
       });
 
-      it('should not refetch groups if main frame URL has not changed', () => {
+      it('should not refetch groups if main frame URL has not changed', async () => {
         const svc = service();
 
         fakeStore.setState({
           frames: [{ uri: 'https://domain.com/page-a' }],
         });
-        return svc
-          .load()
-          .then(() => {
-            return fakeRootScope.eventCallbacks[events.FRAME_CONNECTED]();
-          })
-          .then(() => {
-            assert.calledOnce(fakeApi.groups.list);
-          });
+
+        await svc.load();
+        assert.calledOnce(fakeApi.groups.list);
+
+        // A new frame connects, but the main frame URI remains the same.
+        fakeApi.groups.list.resetHistory();
+        fakeStore.setState({
+          frames: [
+            { uri: 'https://domain.com/page-a' },
+            { uri: 'https://domain.com/iframe-b' },
+          ],
+        });
+
+        await new Promise(resolve => setTimeout(resolve, 1));
```

Comment:
In the test above, we wait until the `GET /api/groups` API call _does_ happen. Here we just wait a short amount of time and make sure that it _doesn't_ happen. We can't just call the assert immediately because there are a few async steps that happen in `groups.load()` before the API call happens.

Reply:
Just want to double check that a `setTimeout` here is preferable to using a `sinon` timer.

Reply:
Good question! Sinon's fake timers can only advance the clock used by `Date` and `setTimeout`. They can't "spin the event loop" and advance code that involves a series of async steps (of unknown length), which is what needs to happen here.